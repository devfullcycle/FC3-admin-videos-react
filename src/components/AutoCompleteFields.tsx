import {
  Autocomplete,
  AutocompleteRenderInputParams,
  TextField,
} from "@mui/material";
import { CastMember } from "../types/CastMembers";
import { Category } from "../types/Category";
import { Genre } from "../types/Genres";

type Props = {
  name: string;
  label: string;
  isLoading: boolean;
  isDisabled: boolean;
  values?: (Genre | Category | CastMember)[];
  options?: (Genre | Category | CastMember)[];
  handleChange: (e: React.ChangeEvent<HTMLInputElement>) => void;
};

export const AutoCompleteFields = ({
  name,
  label,
  values,
  options,
  isLoading,
  isDisabled,
  handleChange,
}: Props) => {
  const valuesById = new Map(options?.map(i => [i.id, i]));

  const renderOptions = (
    props: React.HTMLAttributes<HTMLLIElement>,
    option: Category | Genre | CastMember
  ) => (
    <li {...props} key={option.id}>
      {option.name}
    </li>
  );

  const isEqualId = (
    option: Genre | Category | CastMember,
    value: Genre | Category | CastMember
  ) => {
    return option.id === value.id;
  };

  const handleOnChange = (
    _e: React.ChangeEvent<{}>,
    newValue: (Genre | Category | CastMember)[]
  ) => {
    handleChange({ target: { name, value: newValue } } as any);
  };

  const renderInput = (params: AutocompleteRenderInputParams) => (
    <TextField {...params} label={label} data-testid={`${name}-input`} />
  );

  return (
    <Autocomplete
      multiple
      value={values}
      options={options || []}
      loading={isLoading}
      onChange={handleOnChange}
      renderInput={renderInput}
      data-testid={`${name}-search`}
      renderOption={renderOptions}
      isOptionEqualToValue={isEqualId}
      disabled={isDisabled || !options}
      getOptionLabel={(option) => {
        return valuesById.get(option.id)?.name || ''
      }}
    />
  );
};
