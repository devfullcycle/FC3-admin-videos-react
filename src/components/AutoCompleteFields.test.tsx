import { render } from "@testing-library/react";
import { Category } from "../types/Category";
import { AutoCompleteFields } from "./AutoCompleteFields";

describe("AutoCompleteFields", () => {
  const categories: Category[] = [
    {
      id: "1",
      name: "Comedy",
      deleted_at: "",
      is_active: true,
      created_at: "2022-01-01",
      updated_at: "2022-01-01",
      description: null,
    },
    {
      id: "2",
      name: "Adventure",
      deleted_at: "",
      is_active: true,
      created_at: "2022-01-01",
      updated_at: "2022-01-01",
      description: null,
    },
  ];

  it("should render the component with loading", () => {
    const { asFragment } = render(
      <AutoCompleteFields
        name="field"
        label="Field"
        values={categories}
        options={categories}
        isDisabled={false}
        isLoading={true}
        handleChange={() => {}}
      />
    );
    expect(asFragment()).toMatchSnapshot();
  });
});
