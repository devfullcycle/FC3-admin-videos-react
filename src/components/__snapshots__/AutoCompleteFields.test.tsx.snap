// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AutoCompleteFields should render the component with loading 1`] = `
<DocumentFragment>
  <div
    class="MuiAutocomplete-root MuiAutocomplete-hasClearIcon MuiAutocomplete-hasPopupIcon css-16awh2u-MuiAutocomplete-root"
    data-testid="field-search"
  >
    <div
      class="MuiFormControl-root MuiFormControl-fullWidth MuiTextField-root css-wb57ya-MuiFormControl-root-MuiTextField-root"
      data-testid="field-input"
    >
      <label
        class="MuiFormLabel-root MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-shrink MuiInputLabel-outlined MuiFormLabel-colorPrimary css-1sumxir-MuiFormLabel-root-MuiInputLabel-root"
        data-shrink="true"
        for=":r0:"
        id=":r0:-label"
      >
        Field
      </label>
      <div
        class="MuiInputBase-root MuiOutlinedInput-root MuiInputBase-colorPrimary MuiInputBase-fullWidth MuiInputBase-formControl MuiInputBase-adornedStart MuiInputBase-adornedEnd MuiAutocomplete-inputRoot css-1gywuxd-MuiInputBase-root-MuiOutlinedInput-root"
      >
        <div
          class="MuiButtonBase-root MuiChip-root MuiChip-filled MuiChip-sizeMedium MuiChip-colorDefault MuiChip-deletable MuiChip-deletableColorDefault MuiChip-filledDefault MuiAutocomplete-tag MuiAutocomplete-tagSizeMedium css-iqd1hh-MuiButtonBase-root-MuiChip-root"
          data-tag-index="0"
          role="button"
          tabindex="-1"
        >
          <span
            class="MuiChip-label MuiChip-labelMedium css-6od3lo-MuiChip-label"
          >
            Comedy
          </span>
          <svg
            aria-hidden="true"
            class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium MuiChip-deleteIcon MuiChip-deleteIconMedium MuiChip-deleteIconColorDefault MuiChip-deleteIconFilledColorDefault css-i4bv87-MuiSvgIcon-root"
            data-testid="CancelIcon"
            focusable="false"
            viewBox="0 0 24 24"
          >
            <path
              d="M12 2C6.47 2 2 6.47 2 12s4.47 10 10 10 10-4.47 10-10S17.53 2 12 2zm5 13.59L15.59 17 12 13.41 8.41 17 7 15.59 10.59 12 7 8.41 8.41 7 12 10.59 15.59 7 17 8.41 13.41 12 17 15.59z"
            />
          </svg>
        </div>
        <div
          class="MuiButtonBase-root MuiChip-root MuiChip-filled MuiChip-sizeMedium MuiChip-colorDefault MuiChip-deletable MuiChip-deletableColorDefault MuiChip-filledDefault MuiAutocomplete-tag MuiAutocomplete-tagSizeMedium css-iqd1hh-MuiButtonBase-root-MuiChip-root"
          data-tag-index="1"
          role="button"
          tabindex="-1"
        >
          <span
            class="MuiChip-label MuiChip-labelMedium css-6od3lo-MuiChip-label"
          >
            Adventure
          </span>
          <svg
            aria-hidden="true"
            class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium MuiChip-deleteIcon MuiChip-deleteIconMedium MuiChip-deleteIconColorDefault MuiChip-deleteIconFilledColorDefault css-i4bv87-MuiSvgIcon-root"
            data-testid="CancelIcon"
            focusable="false"
            viewBox="0 0 24 24"
          >
            <path
              d="M12 2C6.47 2 2 6.47 2 12s4.47 10 10 10 10-4.47 10-10S17.53 2 12 2zm5 13.59L15.59 17 12 13.41 8.41 17 7 15.59 10.59 12 7 8.41 8.41 7 12 10.59 15.59 7 17 8.41 13.41 12 17 15.59z"
            />
          </svg>
        </div>
        <input
          aria-autocomplete="list"
          aria-expanded="false"
          aria-invalid="false"
          autocapitalize="none"
          autocomplete="off"
          class="MuiInputBase-input MuiOutlinedInput-input MuiInputBase-inputAdornedStart MuiInputBase-inputAdornedEnd MuiAutocomplete-input MuiAutocomplete-inputFocused css-152mnda-MuiInputBase-input-MuiOutlinedInput-input"
          id=":r0:"
          role="combobox"
          spellcheck="false"
          type="text"
          value=""
        />
        <div
          class="MuiAutocomplete-endAdornment css-1q60rmi-MuiAutocomplete-endAdornment"
        >
          <button
            aria-label="Clear"
            class="MuiButtonBase-root MuiIconButton-root MuiIconButton-sizeMedium MuiAutocomplete-clearIndicator css-1glvl0p-MuiButtonBase-root-MuiIconButton-root-MuiAutocomplete-clearIndicator"
            tabindex="-1"
            title="Clear"
            type="button"
          >
            <svg
              aria-hidden="true"
              class="MuiSvgIcon-root MuiSvgIcon-fontSizeSmall css-ptiqhd-MuiSvgIcon-root"
              data-testid="CloseIcon"
              focusable="false"
              viewBox="0 0 24 24"
            >
              <path
                d="M19 6.41L17.59 5 12 10.59 6.41 5 5 6.41 10.59 12 5 17.59 6.41 19 12 13.41 17.59 19 19 17.59 13.41 12z"
              />
            </svg>
            <span
              class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
            />
          </button>
          <button
            aria-label="Open"
            class="MuiButtonBase-root MuiIconButton-root MuiIconButton-sizeMedium MuiAutocomplete-popupIndicator css-qzbt6i-MuiButtonBase-root-MuiIconButton-root-MuiAutocomplete-popupIndicator"
            tabindex="-1"
            title="Open"
            type="button"
          >
            <svg
              aria-hidden="true"
              class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-i4bv87-MuiSvgIcon-root"
              data-testid="ArrowDropDownIcon"
              focusable="false"
              viewBox="0 0 24 24"
            >
              <path
                d="M7 10l5 5 5-5z"
              />
            </svg>
            <span
              class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
            />
          </button>
        </div>
        <fieldset
          aria-hidden="true"
          class="MuiOutlinedInput-notchedOutline css-1d3z3hw-MuiOutlinedInput-notchedOutline"
        >
          <legend
            class="css-1in441m"
          >
            <span>
              Field
            </span>
          </legend>
        </fieldset>
      </div>
    </div>
  </div>
</DocumentFragment>
`;
